blogd.8.in
.\"
.\" Copyright 2000 Werner Fink, 2000 SuSE GmbH Nuernberg, Germany.
.\" Copyright 2026 SUSE Software Solutions Germany GmbH
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.TH BLOGD 8 "Jul 02, 2026" "Version 2.45" "The SUSE boot concept"
.UC 8
.SH NAME
blogd \- boot logging on /dev/console
.SH SYNOPSIS
.\"
.B /sbin/blogd
.RI [ \-f ]
.SH DESCRIPTION
The
.B blogd
daemon determines the real underlying character device of
.IR /dev/console .
.B blogd
spawns a pty/tty pair to reconnect the current
.I /dev/console
with the slave of the pty/tty pair. During
writing information from this slave to the
real character device a ring buffer is used
to hold the information for writing it to an
.B existing
logging file.
.PP
To fetch the real tty of
.I /dev/console
the program
.BR showconsole (8)
can be used. This has the advantage that
.B blogd
will not hold the real character device of
.I /dev/console
as its controlling tty (would hangup any running
getty on that character device).
.\"
.SH OPTIONS
.TP
.B \-f
This option is used at shutdown, that is reboot, halt, and poweroff
but also for kexec.
.SH BOOT PARAMETERS
The
.B blogd
evaluates the following kernel command line parameters:
.TP
.B blog\&.silent[=1|on|yes|true]
If set, all output to the underlying character devices is suppressed.
This is useful to prevent the console from being flooded with messages.
.TP
.B blog\&.coldboot[=1|on|yes|true]
If set, explicitly enables the early coldstart scan for systemd password 
requests (e.g. for LUKS decryption) during the initrd phase.
.TP
.B blog\&.timeout=<integer>
On 
.B s390x
(64-bit IBM System/390) and
.B z/Architecture
this specifies the timeout for the
.B MORE
prompt on the
.I 3215
console.  If set to
.I 0
the boot process does not need manual support.
.SH SIGNALS
.B blogd
knows a few signals to control its behavior.
.TP
.BR SIGQUIT ", " SIGINT ", and " SIGTERM
will cause
.B blogd
to try to write out the ring buffer and to exit.
.TP
.B SIGIO
tells
.B blogd
that now it is able to write on
.I @@BOOT_LOGFILE@@
which means that the file system is
mounted read/write and the kernel
messages are written to that file.
After a reboot the file
.I @@BOOT_OLDLOGFILE@@
may exist where the former messages
can be found including the messages at
shutdown upto the point where the file
systems become unmounted.
.TP
.B SIGSYS
tells
.B blogd
that it should stop writing to disk but
continue to repeat messages to the old
devices of the system console.
.\"
.SH BUGS
.B blogd
needs a mounted
.I /proc
and
.I /dev/pts
file system and tries to set the controlling
tty to
.B stdin
if the real character device of
.I /dev/console
is not given.  After reading
.I /proc
.B blogd
tries to restore the status of the controlling tty
to avoid problems with getty processes.  This can
fail because
.B blogd
forks to run in the background as a daemon.
.SH FILES
.TP
.I /proc/<pid of blogd>/stat
the stat file of the
.B blogd
process.
.TP
.I /dev/console
the system console.
.TP
.IR @@BOOT_LOGFILE@@ ", " @@BOOT_OLDLOGFILE@@
logging file which is created by
.BR blogd (8)
.\"
.SH SEE ALSO
.BR showconsole (8),
.BR syslogd (8),
.BR proc (5).
.SH COPYRIGHT
2000,2015 Werner Fink,
2000 SuSE GmbH Nuernberg, Germany,
2015 SuSE Linux GmbH,
2026 SUSE Software Solutions Germany GmbH.
.SH AUTHOR
Werner Fink <werner@suse.de>


blogd.c
/*
 * blogd.c
 *
 * Copyright 2000,2015 Werner Fink, 2000 SuSE GmbH Nuernberg, Germany.
 * Copyright 2015 SuSE Linux GmbH.
 *
 * This source is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */

#ifndef  _GNU_SOURCE
# define _GNU_SOURCE
#endif
#include <sys/time.h>
#include <sys/types.h>
#include <sys/sysmacros.h> /* Defines the macros major and minor */
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/vt.h>
#include <sys/kd.h>
#include <time.h>
#include <pty.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <dirent.h>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
#include <limits.h>
#include <errno.h>
#include <paths.h>
#include "libconsole.h"
#ifndef  _POSIX_MAX_CANON
# define _POSIX_MAX_CANON 255
#endif

extern volatile sig_atomic_t nsigsys;
extern volatile sig_atomic_t signaled;
extern volatile sig_atomic_t asking;
extern int final;
extern int console_silent;
extern int coldboot;

static int show_status;
static const char console[] = "/dev/console";
static const char *varrun = _PATH_VARRUN;

static void _initialize(void) __attribute__((__constructor__));
static void _initialize(void)
{
    const char *run;
    run = realpath(varrun, NULL);
    if (run && *run)
	varrun = run;
}

/*
 * Cry and exit.
 */
__attribute__((noreturn)) void error (const char *fmt, ...)
{
    va_list ap;
    va_start(ap, fmt);
    verr(EXIT_FAILURE, fmt, ap);
    va_end(ap);
}

/*
 * Remove pidfile
 */
static const char *myname;
static char *pidfile;
static char *plymouth;
static void rmfpid()
{
    if (!pidfile || *pidfile == 0)
	return;
    unlink(pidfile);
    if (plymouth && *plymouth)
	unlink(plymouth);
}

/*
 * Write pidfile
 */
static void attribute((noinline)) dopidfile()
{
    FILE * fpid;
    int ret;

    ret = asprintf(&pidfile, "%s/%s.pid", varrun, myname);
    if (ret < 0)
	error("can not allocate memory for pid file");

    if ((fpid = fopen (pidfile, "w")) == NULL) {
	warn("can not open %s", pidfile);
	goto out;
    }
    fprintf(fpid, "%d\n", getpid());
    fclose(fpid);

    ret = mkdir("/run/plymouth", 0755);
    if (ret < 0) {
	if (errno != EEXIST)
	    warn("can not make directory /run/plymouth");
	else
	    ret = 0;
    }
    if (ret == 0) {
	plymouth = "/run/plymouth/pid";
	ret = symlink(pidfile, plymouth);
	do {
	    int serr = errno;
	    int fd;

	    if (ret >= 0)
		break;

	    fd = open(plymouth, O_RDONLY|O_NOFOLLOW);
	    if (fd < 0 && errno == ELOOP) {
		    unlink(plymouth);
		    ret = symlink(pidfile, plymouth);
		    break;
	    }
	    if (fd >= 0) {
		char bpid[128];
		ssize_t len;

		len = read(fd, &bpid[0], sizeof(bpid)-1);
		close(fd);
		if (len > 0) {
		    pid_t pid;
		    char *exe;

		    pid = (pid_t) strtol(&bpid[0], &exe, 10);
		    if (pid > 1 && (!exe || !*exe || *exe == '\n')) {
			exe = proc2exe(pid);
			if (exe) {
			    errno = EACCES;
			    error("plymouth is active %ld", (long int)pid);
			}
		    }
		    unlink(plymouth);
		    ret = symlink(pidfile, plymouth);
		    break;
		}
	    }
	    errno = serr;
	    warn("can not create %s", plymouth);
	    plymouth = NULL;

	} while(0);
    }
    atexit(rmfpid);
out:
    return;
}

/*
 *  Signal handler
 */
static struct sigaction saved_sigttin;
static struct sigaction saved_sigttou;
static struct sigaction saved_sigtstp;
static struct sigaction saved_sighup;
static struct sigaction saved_sigpipe;
#ifdef NO_SIGNALFD
static struct sigaction saved_sigint;
static struct sigaction saved_sigquit;
static struct sigaction saved_sigterm;
static struct sigaction saved_sigsys;

static void sighandle(int sig)
{
    if (nsigsys && (sig == SIGTERM))
	return;
    signaled = (volatile sig_atomic_t)sig;
}

/*
 * Stop writing logs to disk, only repeat messages
 */
static void sigsys(int sig)
{
    nsigsys = (volatile sig_atomic_t)sig;
}
#endif

/*
 * To be able to reconnect to real tty on EIO
 */
static int reconnect(int fd)
{
    struct console * c;
    int ret = 0, tflags;
    int olderr = errno;

    list_for_each_entry(c, &lcons, node) {
	int newfd;

	if (!c->tty) continue;

	if (c->fd != fd) continue;

	switch (c->fd) {
	case 0:
	case -1:	/* Weired */
	    break;
	default:	/* IO of system consoles */
	    if ((newfd = open_tty(c->tty, O_WRONLY|O_NONBLOCK|O_NOCTTY|O_CLOEXEC)) < 0)
		error("can not open %s: %m", c->tty);
	    epoll_delete(c->fd);
	    dup2(newfd, c->fd);
	    if (newfd != c->fd)
		close(newfd);
	    epoll_addwrite(c->fd, &epoll_write_watchdog);
	    ret = 1;
#if defined(__s390__) || defined(__s390x__)
	    if (major(c->dev) == 4 && minor(c->dev) == 64)
		break;
#endif
	    if ((tflags = fcntl(c->fd, F_GETFL)) < 0)
		warn("can not get terminal flags of %s", c->tty);
	    tflags &= ~(O_NONBLOCK);
	    if (fcntl(c->fd, F_SETFL, tflags) < 0)
		warn("can not set terminal flags of %s", c->tty);
	    break;
	}
    }
    errno = olderr;

    return ret;
}


static volatile pid_t pid = -1;

static void flush_handler (void) attribute((noinline));
static void exit_handler (void) attribute((noinline));
#if defined(__s390__) || defined(__s390x__)
static void vmcp_handler (void) attribute((noinline));
#endif

/*
 * Now do the job
 */
int main(int argc, char *argv[])
{
    char ptsname[NAME_MAX+1];
    const char *tty, *stt, *val;
    volatile char *arg0;
    struct console *c;
    struct termios o;
    struct winsize w;
    struct stat st;
    time_t tt;
    int ptm, pts, fd, arg;
    int listen, pipefd[2];

    listen = open_un_socket_and_listen();
    if (listen < -1)
	_exit(EXIT_SUCCESS);

    if (stat("/run/systemd/show-status", &st) == 0)
	show_status = 1;
    if (kill (1, SIGRTMIN+20) < 0)
	warn("could not tell system to show its status");

    arg0 = (volatile char*)argv[0];
    remember_arg0(arg0);

    while ((arg = getopt(argc, argv, "f")) != -1) {
	switch (arg) {
	case 'f':
	    final = 1;
	    break;
	case '?':
	default:
	    return 1;
	}
    }
    argv += optind;
    argc -= optind;

    parse_cmdline();	/* Parse the kernel command line for blog.<key>(=<val>) */
    val = value_cmdline("silent");
    if (val) {
        if (strcmp(val, "1") == 0 || strcasecmp(val, "on") == 0 || strcasecmp(val, "yes") == 0 || strcasecmp(val, "true") == 0)
            console_silent = 1;
    }
    val = value_cmdline("coldboot");
    if (val) {
        if (strcmp(val, "1") == 0 || strcasecmp(val, "on") == 0 || strcasecmp(val, "yes") == 0 || strcasecmp(val, "true") == 0)
            coldboot = 1;
    }

    myname = program_invocation_short_name;
    getconsoles(1);

    close(0);
    close(1);
    close(2);
    if ((fd = open_tty(console, O_RDWR|O_NONBLOCK|O_NOCTTY|O_CLOEXEC)) < 0)
	error("Can not open system console %s: %m", console);

    if (fd != 0) {
	(void)ioctl(fd, TIOCNXCL);	/* Avoid EBUSY */
	dup2(fd, 0);
	close(fd);
    } else {
	(void)ioctl(0, TIOCNXCL);
    }

    dup2(0, 1);
    dup2(0, 2);
    tty = console;

    (void)ioctl(0, TIOCCONS, NULL);  /* Undo any current map if any */

    list_for_each_entry(c, &lcons, node) {
	speed_t ospeed;
	speed_t ispeed;
	int flags;

	ispeed = B38400;
	ospeed = B38400;

	if (ioctl(c->fd, TIOCMGET, &flags) == 0) {  /* serial line */
	    ispeed = cfgetispeed(&c->otio);
	    ospeed = cfgetospeed(&c->otio);
	    c->flags |= CON_SERIAL;
	}
	cfsetispeed(&c->otio, ispeed);
	cfsetospeed(&c->otio, ospeed);

	if (c->flags & CON_CONSDEV) {

	    tty = c->tty;

	    w.ws_row = 0;
	    w.ws_col = 0;
	    if (ioctl(c->fd, TIOCGWINSZ, &w) < 0)
		error("can not get window size of %s", c->tty);

	    if (!w.ws_row)
		w.ws_row = 24;
	    if (!w.ws_col)
		w.ws_col = 80;

	    memcpy(&o, &c->otio, sizeof(o));
	    cfmakeraw(&o);
	    cfsetispeed(&o, ispeed);
	    cfsetospeed(&o, ospeed);
	    o.c_lflag &= ~ECHO;
	    o.c_lflag |= ISIG;
	    o.c_cc[VTIME] = 0;
	    o.c_cc[VMIN]  = CMIN;
	}
#if defined(__s390__) || defined(__s390x__)
	if ((c->flags & CON_3215) && final == 0) {
	    char *timeout = value_cmdline("timeout");

	    if (timeout && isinteger(timeout)) {
		int vmcp = openvmcp();
		if (vmcp >=0) {
		    char *msg = queryterm(vmcp);
		    if (msg) {
			parseterm(msg);
			free(msg);
			setterm(vmcp, timeout);
		    }
		    close(vmcp);
		    atexit(vmcp_handler);	/* Register vmcp restore exit handler */
		}
		ioctl(c->fd, TIOCCBRK);
		usleep(1000);
	    }

	}
#endif
    }

    free_cmdline();	/* free cmd list after last usage of value_cmdline() */

    if (openpty(&ptm, &pts, ptsname, &o, &w) < 0)
	error("can not open pty/tty pair");

    if (fstat(pts, &st) < 0)
	error("can not stat slave pty");
    else {
	struct termios lock;
	memset(&lock, 0xff, sizeof(lock));
	(void)ioctl(pts, TIOCSLCKTRMIOS, &lock);
    }

    if (ioctl(pts, TIOCCONS, NULL) < 0)
	error("can not set console device to %s", ptsname);

    /*
     * Reconnecting to stdin aka ptm to 0 is done after we fork away
     */
#ifdef DEBUG
    int l;
    if ((l = open("/dev/shm/blog.out", O_WRONLY|O_NOCTTY|O_NONBLOCK|O_CREAT|O_APPEND, S_IWUSR|S_IRUSR|S_IRGRP|S_IROTH)) >= 0) {
	dup2(l, STDOUT_FILENO);
	dup2(l, STDERR_FILENO);
	close(l);
    } else
#endif
    {
	dup2(pts,  1);
	dup2(pts,  2);	/* Now we are blind upto safeIO() loop */
    }
    if (pts > 2)
	close(pts);

    asking = signaled = nsigsys = 0;
    set_signal(SIGTTIN, &saved_sigttin, SIG_IGN);
    set_signal(SIGTTOU, &saved_sigttou, SIG_IGN);
    set_signal(SIGTSTP, &saved_sigtstp, SIG_IGN);
    set_signal(SIGHUP,  &saved_sighup,  SIG_IGN);
    set_signal(SIGPIPE, &saved_sigpipe, SIG_IGN);
#ifdef NO_SIGNALFD
    set_signal(SIGINT,  &saved_sigint,  sighandle);
    set_signal(SIGQUIT, &saved_sigquit, sighandle);
    set_signal(SIGTERM, &saved_sigterm, sighandle);
    set_signal(SIGSYS,  &saved_sigsys,  sigsys);
    (void)restart_sig(SIGINT,  1);
    (void)restart_sig(SIGQUIT, 1);
    (void)restart_sig(SIGTERM, 1);
    (void)restart_sig(SIGSYS,  1);
#endif

    list_for_each_entry(c, &lcons, node) {
	struct termios oldtio;
	speed_t ospeed;
	speed_t ispeed;
	int flags;

	(void)ioctl(c->fd, TIOCNXCL);	/* Avoid EBUSY */

#ifdef _PC_MAX_CANON
	if ((c->max_canon = (ssize_t)fpathconf(c->fd, _PC_MAX_CANON)) <= 0)
#endif
	    c->max_canon = _POSIX_MAX_CANON;
	c->tlock = 0;
	if (tcgetattr(c->fd, &c->otio) < 0)
	    continue;
	c->tlock = 1;

	if (ioctl(c->fd, TIOCGLCKTRMIOS, &c->ltio) == 0) {
	    /*
	     * Remove any lock as e.g. startpar(8) will use it
	     */
	    struct termios lock;
	    memset(&lock, 0x00, sizeof(lock));
	    if (ioctl(c->fd, TIOCSLCKTRMIOS, &lock) == 0)
		c->tlock = 2;
	}

	oldtio = c->otio;				/* Remember old tty I/O flags */

	if (ioctl(c->fd, TIOCMGET, &flags) == 0) {	/* serial line */
	    ispeed = cfgetispeed(&c->otio);
	    ospeed = cfgetospeed(&c->otio);

	    c->otio.c_iflag  = c->otio.c_lflag = 0;
	    c->otio.c_oflag  = (ONLCR | OPOST);
	    c->otio.c_oflag &= ~(OLCUC);
	    c->otio.c_cflag  = CREAD | CS8 | HUPCL | (c->otio.c_cflag & CLOCAL);

	    cfsetispeed(&c->otio, ispeed);
	    cfsetospeed(&c->otio, ospeed);
	} else {
	    if (major(c->dev) == 4 && minor(c->dev) <= 63)
		ioctl(c->fd, KDSETMODE, KD_TEXT);	/* Enforce text mode */

	    c->otio.c_iflag |= (ICRNL | IXON);
	    c->otio.c_iflag &= ~(INLCR | IGNCR | BRKINT);
	    c->otio.c_oflag |= (ONLCR | OPOST);
	    c->otio.c_oflag &= ~(OCRNL | ONLRET | OLCUC);
	}

	(void)tcsetattr(c->fd, TCSADRAIN, &c->otio);

	c->ctio = c->otio;				/* Current setting */
	c->otio = oldtio;				/* To be able to restore */
    }

    atexit(exit_handler);	/* Register main exit handler */

    /* Make the first byte in argv be '@' so that we can survive systemd's killing
     * spree when going from initrd to /, and so we stay alive all the way until
     * the power is killed at shutdown.
     * http://www.freedesktop.org/wiki/Software/systemd/RootStorageDaemons
     */
    if (access("/etc/initrd-release", F_OK) >= 0 || final)
	arg0[0] = '@';

    if (pipe2(pipefd, O_CLOEXEC) < 0)
	pipefd[0] = pipefd[1] = -1;

    switch ((pid = fork())) {
    case 0:
#ifdef DEBUG
{
    int l;
    if ((l = open("/dev/shm/blog.out", O_WRONLY|O_NOCTTY|O_NONBLOCK|O_CREAT|O_APPEND, S_IWUSR|S_IRUSR|S_IRGRP|S_IROTH)) >= 0) {
	dup2(l, STDOUT_FILENO);
	dup2(l, STDERR_FILENO);
	close(l);
    }
}
#endif
	/* Write pid file */
	dopidfile();
	/* Get our own session */
	setsid();
	/* Reconnect our own terminal I/O */
	dup2(ptm, 0);
	{
	    const char *msg = "can not get terminal flags of stdin\n";
	    int flags;
	    int ret __attribute__ ((unused));

	    if ((flags = fcntl(0, F_GETFL)) < 0)
		list_for_each_entry(c, &lcons, node) {
		    if (c->fd < 0)
			continue;
		    ret = write(c->fd, msg, strlen(msg));
		}
	    else {
		flags &= ~(O_NONBLOCK);
		flags |=   O_NOCTTY;
		if (fcntl(0, F_SETFL, flags) < 0)
		    list_for_each_entry(c, &lcons, node) {
			if (c->fd < 0)
			    continue;
			ret = write(c->fd, msg, strlen(msg));
		    }
	    }
	}
	if (ptm > 0)
	    close(ptm);
	if (pipefd[0] >= 0)
	    close(pipefd[0]);
	break;
    case -1:
	if (ptm > 0)
	    close(ptm);
	if (pipefd[0] >= 0)
	    close(pipefd[0]);
	if (pipefd[1] >= 0)
	    close(pipefd[1]);
	list_for_each_entry(c, &lcons, node) {
	    const char *msg = "blogd: can not fork to become daemon: ";
	    const char *err = strerror(errno);
	    int ret __attribute__ ((unused));
	    if (c->fd < 0)
		continue;
	    ret = write(c->fd, msg, strlen(msg));
	    ret = write(c->fd, err, strlen(err));
	    ret = write(c->fd, "\n", 1);
	}
	exit(EXIT_FAILURE);
    default:
	time(&tt);
	stt = ctime(&tt);
	close(ptm);
	if (pipefd[1] >= 0)
	    close(pipefd[1]);
	list_for_each_entry(c, &lcons, node) {
	    if (c->fd > 0) {
		close(c->fd);
		c->fd = -1;
	    }
	}
	fflush(stdout);
	fprintf(stdout, "\rBoot logging started on %s(%s) at %.24s\n", tty, console, stt);
	fflush(stdout);

	if (pipefd[0] > 0) {
	    int dummy, ret __attribute__ ((unused));
	    ret = read(pipefd[0], &dummy, 1);
	    close(pipefd[0]);
	}

	_exit(EXIT_SUCCESS);	/* NEVER rise exit handlers here */
    }

    atexit(flush_handler);	/* Register flush exit handler */

    prepareIO(reconnect, listen, 0);

    if (pipefd[1] >= 0)
	close(pipefd[1]);

    while (!signaled)
	safeIO();

    exit(EXIT_SUCCESS);		/* Raise exit handlers */
}

static void flush_handler (void)
{
    (void)tcdrain(1);
    (void)tcdrain(2);
    closeIO();
}

static void exit_handler (void)
{
    struct console *c;
    int fd;

    if (show_status == 0 && kill (1, SIGRTMIN+20) < 0)
	warn("could not tell system to hode its status");

    list_for_each_entry(c, &lcons, node) {
	if (c->fd > 0) {
	    if (c->tlock)	/* write back old setup  */
		tcsetattr(c->fd, TCSADRAIN, &c->otio);
	    if (c->tlock > 1)	/* write back lock if any */
		(void)ioctl(c->fd, TIOCSLCKTRMIOS, &c->ltio);
	    c->tlock = 0;
	    if (c->fd > 1)
		close(c->fd);
	    c->fd = -1;
	}
    }

    errno = 0;
    if ((fd = open_tty(console, O_RDWR|O_NOCTTY|O_CLOEXEC|O_NONBLOCK)) >= 0) {
	int flags;

	flags = fcntl(fd, F_GETFL);
	flags &= ~O_NONBLOCK;
	fcntl(fd, F_SETFL, flags);

	(void)ioctl(fd, TIOCCONS, NULL);	/* Restore old console mapping */
	if (fd > 0)
	    close(fd);
    }

    close(1);
    close(2);
    (void)tcflush(0, TCIOFLUSH);
    close(0);

    reset_signal(SIGTTIN, &saved_sigttin);
    reset_signal(SIGTTOU, &saved_sigttou);
    reset_signal(SIGTSTP, &saved_sigtstp);
    reset_signal(SIGHUP,  &saved_sighup);
    reset_signal(SIGPIPE, &saved_sigpipe);
#ifdef NO_SIGNALFD
    reset_signal(SIGINT,  &saved_sigint);
    reset_signal(SIGQUIT, &saved_sigquit);
    reset_signal(SIGTERM, &saved_sigterm);
    reset_signal(SIGSYS,  &saved_sigsys);
#endif
}

#if defined(__s390__) || defined(__s390x__)
static void vmcp_handler (void)
{
    int vmcp = openvmcp();
    if (vmcp >= 0) {
	restoreterm(vmcp);
	restorespool(vmcp);
	clearvmcp();
	close(vmcp);
    }
}
#endif
